specs/DesignDiscussions/InputValidation.md

```diff
@@ -0,0 +1,260 @@
+# Input Validation
+
+Input validation was originally spec'd in issue [#3081](https://github.com/microsoft/AdaptiveCards/issues/3081). This document covers the basics from that spec, and continues the discussion of additional concerns that it raised.
+
+## Schema Additions to Inputs for Validation
+
+``` json
+{
+    "type": "Input.Text",
+    "id": "username",
+    "isRequired": true,
+    "regex": "^\\w{5,50}$",
+    "errorMessage": "Username must be between 5 and 50 characters long."
+}
+```
+The main additions to the schema for input validation are the `"isRequired"`, `"regex"`, and `"errorMessage"` properties on inputs. These properties allow a card author to specify whether or not an input is required, what regular expression defines valid input, and what error message should be shown in case of invalid user input.
+
+Although `regex` would only be supported on `Input.Text`, the `isRequired` and `errorMessage` properties would be supported on all `Input.*` types.
+
+## Schema Additions to Adaptive Card for Validation
+
+In order to allow the card author to visually indicate whether inputs in a card are required or optional, we will add a new property, `showRequiredInputHints`. When this is true, visual indicators will be added to required inputs inputs to signal to the user that they are required. This property will default to true.
+
+```json
+{
+	"type": "AdaptiveCard",
+	"version": "1.3",
+	"showRequiredInputHints":true,
```

Comment:
Is this needed? 

Discussion: 
The host could do this manually by adding their own *.
Semantic information is good.
Do we want host styles for how to style required and options inputs?

We don't need this. Card author can do it today, or we may want to add host features for this in the future.

Reply:
On labels: We need to balance between host control and card author control. Current spec calls for card author control.

---
In reply to: [355022762](https://github.com/microsoft/AdaptiveCards/pull/3663#discussion_r355022762) [](ancestors = 355022762)

Reply:
If there are not labels (the card author is using placeholders), we may still need hints.

---
In reply to: [355023402](https://github.com/microsoft/AdaptiveCards/pull/3663#discussion_r355023402) [](ancestors = 355023402,355022762)

Reply:
Further discussion of labels today: 
- All agreed that P0 base behavior is a plain text label on the input, required inputs marked by a *.
- Next most important thing is allowing host control. Two dimensions: styling, and required indicator string and position
- David feels strongly that all styling should go through CSS (or equivalent) and should not be in host config.
- Matt and i feel some base host config control may be useful. Discussion of underlying use and usefulness of host config in general
- All agreed that native styling should be possible for required vs optional.
- On string (*) configuration, we could use the common expression language. David is concerned about taking a dependency on it for size reasons.
- If host can configure the string, they need to be able to localize it.
- David thinks that allowing hosts to change the string (without loc) should be high priority.

---
In reply to: [355029970](https://github.com/microsoft/AdaptiveCards/pull/3663#discussion_r355029970) [](ancestors = 355029970,355023402,355022762)

Reply:
All of this discussion is out of the scope of this document, and needs to be reflected in the labels spec. I'll make an update to that soon.

---
In reply to: [367159737](https://github.com/microsoft/AdaptiveCards/pull/3663#discussion_r367159737) [](ancestors = 367159737,355029970,355023402,355022762)

Reply:
We've got a discussion scheduled to talk about host config and defaults in general. Do you have a specific concern related to this spec?

---
In reply to: [367215908](https://github.com/microsoft/AdaptiveCards/pull/3663#discussion_r367215908) [](ancestors = 367215908)

Reply:
Added updates to the label spec to this PR to cover the current state of this discussion.

---
In reply to: [373689342](https://github.com/microsoft/AdaptiveCards/pull/3663#discussion_r373689342) [](ancestors = 373689342,367159737,355029970,355023402,355022762)

Reply:
TODO: identify what host config will allow and make proposals on how would this be done